[dtensor] add util to compute expected local sizes/strides for even sharding

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164437
* #164433
* #164431
* #163602
* #164340
* #164420
* #164419
* #164321
* __->__ #164296

Reviewed GPT5 summary:

**Summary / Goal**  
Add a utility to compute expected local tensor sizes and strides under *even sharding* in dtensor.

**Details**  
- New function in `torch/distributed/tensor/_utils.py`.  
- Computes local sizes/strides given global shape, mesh, and placements.  
- Enforces divisibility of global dimension by mesh size (strict even sharding).  
- Complements `compute_global_tensor_info`.  

**Motivation**  
Ensures correctness for stride/layout computations in distributed tensors.  

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

xmfan committed:
[dtensor] add util to compute expected local sizes/strides for even sharding

[ghstack-poisoned]

zpcore commented on torch/distributed/tensor/_utils.py:
```diff
@@ -271,6 +271,70 @@ def compute_global_tensor_info(
+    for idx, placement in enumerate(placements):
+        mesh_dim_size = mesh.size(idx)
+        if placement.is_shard():
+            shard_placement = cast(Shard, placement)
```
I think this function doesn't apply for `_StridedShard` right? I think it is fine here since it will be deprecated.

zpcore commented on torch/distributed/tensor/_utils.py:
```diff
@@ -271,6 +271,70 @@ def compute_global_tensor_info(
+            )
+
+            global_dim_size = local_shape[shard_dim]
+            assert global_dim_size % mesh_dim_size == 0, (
```
Why do we need `global_dim_size % mesh_dim_size == 0` here?

ezyang commented:
There is a little bit of a smell here, in the sense that, ostensibly there was equivalent logic for this for the uneven sharding case, but now you have a completely disjoint implementation, why does it have to be disjoint?

xmfan commented:
To support uneven shapes like in `compute_global_tensor_shape`, you need to allgather all local shards. If you know you have even sharding, you don't want to launch that collective. Can I fast follow here? the entire stack depends on this.

ezyang commented on torch/distributed/tensor/_utils.py:
```diff
@@ -271,6 +271,70 @@ def compute_global_tensor_info(
+            assert shard_dim < len(local_shape), (
+                f"Sharding dim {shard_dim} greater than tensor ndim {len(local_shape)} "
+                f"for placement number {idx}."
+            )
```
These asserts are written in a kind of weirdly stilted way. LLM?

ezyang replied:
Oh it's actually lifted straight from `compute_global_tensor_info`

ezyang commented on torch/distributed/tensor/_utils.py:
```diff
@@ -271,6 +271,70 @@ def compute_global_tensor_info(
+
+    For example, if we have a global tensor with size (4, 8, 4) and stride (32, 1, 8).
+    If the DTensor placements are [Shard(2)] and world_size is 2;
+    then the local size is (4, 8, 2) and stride is (16, 1, 8).
```
It seems worth pointing out this only works for even sharding. Because its companion compute_global_tensor_info obviously only works for even sharding (since you only have the local info), but this function COULD in principle work for uneven.

ezyang commented on torch/distributed/tensor/_utils.py:
```diff
@@ -271,6 +271,70 @@ def compute_global_tensor_info(
+                    i != shard_dim
+                    and local_stride[i] >= local_stride[shard_dim] * mesh_dim_size
+                ):
+                    local_stride[i] = local_stride[i] // mesh_dim_size
```
This is guaranteed to be exactly divisible too?

xmfan replied:
for contiguous tensors at least. i'm not sure we need to handle arbitrary strides? it's not clear to me how any of the other utils handle those

ezyang reviewed (approved):
stamping to make sure the stack isn't blocked

xmfan commented:
`compute_local_shape_and_global_offset` should return the same shape, but does not return strides. it instead returns the global offset / coordinate / n-d jax axis_dim

`compute_local_stride` should return the same strides, but only computes the strides

xmfan committed:
Update on "[dtensor] add util to compute expected local sizes/strides for even sharding"


Reviewed GPT5 summary:

**Summary / Goal**  
Add a utility to compute expected local tensor sizes and strides under *even sharding* in dtensor.

**Details**  
- New function in `torch/distributed/tensor/_utils.py`.  
- Computes local sizes/strides given global shape, mesh, and placements.  
- Enforces divisibility of global dimension by mesh size (strict even sharding).  
- Complements `compute_global_tensor_info`.  

**Motivation**  
Ensures correctness for stride/layout computations in distributed tensors.  

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
